### PR TITLE
Add integration tests for kube-monitoring components

### PR DESCRIPTION
**Pull request by ibakshay**
This pull request adds integration tests for the kube-monitoring components. The tests include checking the connectivity to the kube-monitoring-prometheus, kube-monitoring-state-metrics, and kube-monitoring-node-exporter services. The tests also verify the setup of the kube-monitoring-operator components, the creation of the persistent volume, the presence of the kube-monitoring resource of type 'Prometheus', and the creation of the required CRDs for kube-monitoring.

Fixes cloudoperators/greenhouse-extensions#217
deponds on https://github.com/cloudoperators/greenhouse-extensions/pull/237

### CODE REVIEW

File: kube-monitoring/charts/templates/tests/test-permissions.yaml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    "helm.sh/hook": test
+    "helm.sh/hook-weight": "-5" # this annotation is used so that this serviceaccount is created before the test pod gets executed
+    "helm.sh/hook-delete-policy": "before-hook-creation,hook-succeeded" # this annotation is used to delete the test pod after the test is successful
+---
+
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  namespace: default
+  name: object-reader
```
Comment (richardtief) [inline kube-monitoring/charts/templates/tests/test-permissions.yaml:17]:
```suggestion
  name: {{ .Release.Name }}-test
```

File: kube-monitoring/charts/templates/tests/test-permissions.yaml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    verbs: ["get", "list"]
+---
+
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: crd-viewer-binding
+subjects:
+  - kind: ServiceAccount
+    name: integration-test
```
Comment (richardtief) [inline kube-monitoring/charts/templates/tests/test-permissions.yaml:54]:
```suggestion
    name: {{ .Release.Name }}-test
```

**Review by richardtief — CHANGES_REQUESTED** [review]
Marvelous, I really like it! I've left a few comments. Especially with the naming of the resources, there can be potential problems later if you install the kube-monitoring plugin multiple times, e.g. to get just one more Prometheus. 

I have seen that your Git client is not handling newlines correctly. You can find some hints here: https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings

File: kube-monitoring/charts/templates/tests/test-kube-monitoring-config.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
+{{- if .Values.testFramework.enabled -}}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test-kube-monitoring-config
+  labels:
+    type: integration-test
+  annotations:
+    "helm.sh/hook": test
```
[the remark this answers is not in the document]
Comment (ibakshay) [reply kube-monitoring/charts/templates/tests/test-kube-monitoring-config.yaml:9]: Very good finding. We only need this annotation for the pods that are running the integration tests. We do not need this annotation for other manifests like `configmaps`, `serviceaccounts`, `rolebinding` and so on. These resources will be already created during the `helm install/upgrade <deployment name>` command if these annotations are not present.

[the remark this answers is not in the document]
Comment (ibakshay) [reply kube-monitoring/charts/templates/tests/test-kube-monitoring-config.yaml:9]: The resources without the `helm.sg/hook: test` annotation  won't be cleaned up after the `helm test` has been executed. Is that fine from your perspective?

File: kube-monitoring/charts/templates/tests/test-kube-monitoring-config.yaml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    "helm.sh/hook-weight": "-5"
+    "helm.sh/hook-delete-policy": "before-hook-creation,hook-succeeded" # this annotation is used to delete the test pod after the test is successful
```
[the remark this answers is not in the document]
Comment (ibakshay) [reply kube-monitoring/charts/templates/tests/test-kube-monitoring-config.yaml]: I have removed the comments in https://github.com/cloudoperators/greenhouse-extensions/pull/236/commits/8dba9e8ff4e32104df23caebd69630060e5e2d6a commit.

File: kube-monitoring/charts/templates/tests/test-kube-monitoring.yaml [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{{- if .Values.testFramework.enabled -}}
+apiVersion: v1
+kind: Pod
+metadata: 
+  name: kube-monitoring-integration-test
+  labels: 
+    type: integration-test-test
```
[the remark this answers is not in the document]
Comment (ibakshay) [reply kube-monitoring/charts/templates/tests/test-kube-monitoring.yaml]: great insight. I have added the common lablels to every resource in this commit https://github.com/cloudoperators/greenhouse-extensions/pull/236/commits/8dba9e8ff4e32104df23caebd69630060e5e2d6a.

File: kube-monitoring/charts/templates/tests/test-kube-monitoring.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  annotations:
+    "helm.sh/hook": test
+    "helm.sh/hook-delete-policy": "before-hook-creation,hook-succeeded" # this annotation is used to delete the test pod after the test is successful
+spec:
+  serviceAccountName: integration-test
+  containers: 
+    - name: bats-test
```
[the remark this answers is not in the document]
Comment (ibakshay) [reply kube-monitoring/charts/templates/tests/test-kube-monitoring.yaml:15]: fixed in this commit https://github.com/cloudoperators/greenhouse-extensions/pull/236/commits/8dba9e8ff4e32104df23caebd69630060e5e2d6a.

File: kube-monitoring/charts/templates/tests/test-kube-monitoring.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ibakshay) [reply kube-monitoring/charts/templates/tests/test-kube-monitoring.yaml]: fixed in this commit https://github.com/cloudoperators/greenhouse-extensions/pull/236/commits/8dba9e8ff4e32104df23caebd69630060e5e2d6a.

File: kube-monitoring/charts/templates/tests/test-kube-monitoring-config.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (richardtief) [reply kube-monitoring/charts/templates/tests/test-kube-monitoring-config.yaml:10]: Ok, then I would leave it as it is. It's even worse if the permissions are lying around, which is not the best security we can offer.

File: kube-monitoring/charts/templates/tests/test-permissions.yaml [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    "helm.sh/hook-delete-policy": "before-hook-creation,hook-succeeded"
+subjects:
+  - kind: ServiceAccount
+    name: {{ .Release.Name }}-test
+    namespace: {{ .Release.Namespace }}
+roleRef:
+    kind: ClusterRole
+    name: {{ .Release.Name }}-test
+    apiGroup: rbac.authorization.k8s.io
+---
```
Comment (richardtief) [inline kube-monitoring/charts/templates/tests/test-permissions.yaml:94]: The last three hyphens are not required.
```suggestion
```

**Review by richardtief — CHANGES_REQUESTED** [review]
Two nits before we merge: 

1. newlines are still not added at the end of some files. Git needs them to render the diffs correctly. Could you please add them?
2. commented below.

**Comment by ibakshay** [comment]
> Two nits before we merge:
> 
> 1. newlines are still not added at the end of some files. Git needs them to render the diffs correctly. Could you please add them? To always have them: https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings
> 2. commented below.

I have configured by text editor to add new lines at the end of file automatically now :) Thank you for the hint!

**Review by richardtief — APPROVED** [review]
Thanks, let's ship it!

**Pull request merged**